realm/realm-library/src/main/java/io/realm/internal/OsSet.java

```diff
@@ -90,15 +106,99 @@ public boolean remove(@Nullable Object value) {
+    }
+
+    public boolean isSubSetOf(long otherSetNativePtr) {
+        return nativeIsSubSetOf(this.nativePtr, otherSetNativePtr);
+    }
+
+    public boolean containsAll(Collection<?> collection, Class<?> valueClass) {
+        Object[] objects = collection.toArray();
+
+        // It cannot be contained if it is empty or not the same type
+        if (collection.size() == 0 || objects[0].getClass() != valueClass) {
+            return false;
+        }
+
+        if (valueClass == String.class) {
+            String[] values = collection.toArray(new String[objects.length]);
+            return nativeContainsAllString(nativePtr, values);
+        } else {
+            // TODO: add missing types
+            throw new UnsupportedOperationException("set containsAll - Hold your horses cowboy...");
+        }
+    }
+
+    public boolean union(OsSet otherRealmSet) {
+        return nativeUnion(nativePtr, otherRealmSet.getNativePtr());
+    }
+
+    public <E> boolean addAll(Collection<? extends E> collection, Class<?> valueClass) {
+        return collectionFunnel(collection, valueClass, ExternalCollectionOperation.ADD_ALL);
+    }
+
+    public boolean asymmetricDifference(OsSet otherSet) {
+        return nativeAsymmetricDifference(nativePtr, otherSet.getNativePtr());
+    }
+
+    public <E> boolean removeAll(Collection<? extends E> collection, Class<?> valueClass) {
+        return collectionFunnel(collection, valueClass, ExternalCollectionOperation.REMOVE_ALL);
+    }
+
+    public boolean intersect(OsSet otherSet) {
+        return nativeIntersect(nativePtr, otherSet.getNativePtr());
+    }
+
+    public <E> boolean retainAll(Collection<? extends E> collection, Class<?> valueClass) {
+        return collectionFunnel(collection, valueClass, ExternalCollectionOperation.RETAIN_ALL);
     }
 
     public void clear() {
         nativeClear(nativePtr);
     }
 
+    public OsSet freeze(OsSharedRealm frozenSharedRealm) {
+        long frozenNativePtr = nativeFreeze(this.nativePtr, frozenSharedRealm.getNativePtr());
+        return new OsSet(frozenSharedRealm, frozenNativePtr);
+    }
+
+    private <E> boolean collectionFunnel(Collection<? extends E> collection,
+                                         Class<?> valueClass,
+                                         ExternalCollectionOperation operation) {
+        // Return unchanged if collection is empty
+        if (collection.size() == 0) {
+            return false;
+        }
+
+        // Collection cannot be removed if it is not the same type
+        Object[] objects = collection.toArray();
+        Class<?> itemClass = objects[0].getClass();
+        if (itemClass != valueClass) {
+            throw new IllegalArgumentException("Invalid collection type. Set and collection must contain the same type of elements.");
+        }
+
+        if (itemClass == String.class) {
+            String[] values = collection.toArray(new String[objects.length]);
+            switch (operation) {
```

Comment:
If you get any idea on an appropriate API lower down the stack to avoid this, please send it my way, and I will try to ensure that we don't need this _cartesian product_-implementation for Kotlin.